### DISCUSSION TITLE
Replicating tupledict in Gurobi when initializing cp.Variable for efficient computing

### DISCUSSION THREAD

**Discussion by alexandregranzerguay** [cvxpy/cvxpy#2471] [Q&A] [answered]
Hi,

I am looking for advice on how to efficiently handle subsetting and indexing in CVXPY.
Example:

I have a variable x indexed from 0 to 9, and I want to divide this variable into two subsets:

    x[0:4] belongs to subset x1
    x[5:9]belongs to subset x2

I want to formulate and solve the following optimization problem:
min⁡ xTQx
Subject to:
x1⋅A1≤b1
x2⋅A2≤b2

My issue:
I can solve this by looping over the relevant indices, but this approach loses the benefits of vectorization, which is important for the performance of my application. Alternatively, I can define x1 and x2 as separate variables, but then I lose the initial indexing from 0 to 9, which is crucial because I am working with a classification tree where the indices must remain consistent.

My question:
Is there a way to define subsets of a variable in CVXPY while maintaining the original indexing and leveraging vectorization for efficiency? Any suggestions or best practices would be greatly appreciated.

Thank you!

**Comment by SteveDiamond** [comment]
I believe any sort of subsetting can be handled through indexing by boolean arrays. But if that's not the case for your problem I would be curious to hear more details.

**Reply by alexandregranzerguay** [reply]
I gave a bit more clarification in response to @phschiele

**Comment by phschiele** [comment, marked as answer]
You can do exactly what you showed.

```py
x = cp.Variable(10)

x1 = x[0:5]
x2 = x[5:10]  # Note that the right bound is exclusive  

constraints = [
    A1 @ x1 <= b1,
    A2 @ x2 <= b2
]
```

**Reply by alexandregranzerguay** [reply]
Totally agree. The part that gets tricky is the following. Say I have an activation variable z defined only for the x2 set. Can I avoid create a size 10 variable and only use the indexes [5:10].

I suppose the recommendation here is to do the index conversion post solve?

```python
x = cp.Variable(10)
z = cp.Variable(5, boolean=True)

x1 = x[0:5]
x2 = x[5:10]  # Note that the right bound is exclusive  

constraints = [
    A1 @ x1 <= b1,
    A2 @ x2 <= b2,
    x2.T @ z == 1,
]
true_index = np.where(z.value == 1)[0][0] + 5
``` 

This is a fine approach in principle but it can get slightly convoluted when the occurs for many instances across a number of subsets

**Reply by phschiele** [reply]
Yes, that would work! 
As you say, you can either create `x1` and `x2` as separate variables, or use slice in the larger variable. Looks like you need to do some bookkeeping in either case. Both should be fine in terms of performance, which was one of your concerns with looping.

**Reply by alexandregranzerguay** [reply]
Right. Got it.
And if I wanted to index everything from 1 instead of 0, there is no native way to do this correct?

**Reply by phschiele** [reply]
Yes, I'm afraid that's something you need to adapt your code to.
